pallets/gear/src/ext.rs

```diff
@@ -235,15 +234,24 @@ impl EnvExt for LazyPagesExt {
 
         let page_number = self.inner.return_and_store_err(result)?;
 
-        // Add new allocations to lazy pages
+        // Add new allocations to lazy pages.
+        // All pages except ones which has been already allocated,
+        // during current execution.
+        // This is because only such pages contains Default (zeros in web asm) page data.
+        // Pages which has been already allocated may contain garbage.
         let id = self.inner.program_id.into_origin();
-        let new_allocations = self.inner.allocations_context.allocations();
-        for page in new_allocations
-            .difference(&old_allocations)
-            .flat_map(|p| p.to_gear_pages_iter())
-        {
-            log::debug!("add {:?} to lazy pages", page);
-            save_page_lazy_info(id, page);
+        for page_num in page_number.0..(page_number + pages_num).0 {
```

Comment:
would be nice to see some explicit names?

Reply:
Done